pkg/apis/apps/kfdef/v1beta1/application_types.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1beta1
+
+import (
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	KfConfigFile = "app.yaml"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// KfDef is the Schema for the applications API
+// +k8s:openapi-gen=true
+type KfDef struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   KfDefSpec   `json:"spec,omitempty"`
+	Status KfDefStatus `json:"status,omitempty"`
+}
+
+type KfDefSpec struct {
+	Applications []Application `json:"applications,omitempty"`
+	Plugins      []Plugin      `json:"plugins,omitempty"`
+	Secrets      []Secret      `json:"secrets,omitempty"`
+	Repos        []Repo        `json:"repos,omitempty"`
+}
+
+// Application defines an application to install
+type Application struct {
+	Name            string           `json:"name,omitempty"`
+	KustomizeConfig *KustomizeConfig `json:"kustomizeConfig,omitempty"`
+}
+
+type KustomizeConfig struct {
+	RepoRef    *RepoRef    `json:"repoRef,omitempty"`
+	Overlays   []string    `json:"overlays,omitempty"`
+	Parameters []NameValue `json:"parameters,omitempty"`
+}
+
+type RepoRef struct {
+	Name string `json:"name,omitempty"`
+	Path string `json:"path,omitempty"`
+}
+
+type NameValue struct {
+	Name         string `json:"name,omitempty"`
+	Value        string `json:"value,omitempty"`
+	InitRequired bool   `json:"initRequired,omitempty"`
+}
+
+// Plugin can be used to customize the generation and deployment of Kubeflow
+type Plugin struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec *runtime.Object `json:"spec,omitempty"`
+}
+
+// Secret provides information about secrets needed to configure Kubeflow.
+// Secrets can be provided via references e.g. a URI so that they won't
```

Comment:
In the SecretSource struct, I don't see a way to provide a secret with a URI.
However, I do think it's a valid use case.
Maybe we could add a repoSource which will be a RepoRef struct and find a secret from one of the repos? 

cc @jlewi

Reply:
We can always add fields later. How about waiting to add a URI field until we have a concrete use case?

Secrets generally shouldn't be stored in repositories so why would we need the ability to reference a secret in a repository?

Reply:
In that case, the comment should be updated to reflect that.

Reply:
removed URI parts.